File: packages/fallow-agent/src/fallow_agent/modelcache/store.py
"""``HttpModelStore``: the concrete agent-side model cache.

Implements :class:`fallow_protocol.interfaces.ModelStore`. Pulls GGUF blobs
from the coordinator with resume + retry, verifies sha256/size, and keeps
verified files on local disk keyed by a cheap marker so the heartbeat hot path
never rehashes a multi-GB file.
"""

import asyncio
from collections.abc import Awaitable, Callable
from pathlib import Path

import httpx

from fallow_agent.modelcache.config import (
    BLOB_PATH_TEMPLATE,
    DEFAULT_CACHE_DIR,
    MARKER_SUFFIX,
    MMPROJ_PATH_TEMPLATE,
    PART_SUFFIX,
    ModelCacheConfig,
)
from fallow_agent.modelcache.download import (
    DownloadResult,
    RetryableStatusError,
    stream_to_part,
)
from fallow_agent.modelcache.errors import ModelFetchError, ModelVerificationError
from fallow_agent.modelcache.paths import (
    blob_path,
    marker_path,
    read_marker,
    write_marker_atomic,
)
from fallow_protocol.interfaces import ModelStore
from fallow_protocol.models import ModelManifest

SleepFn = Callable[[float], Awaitable[None]]


def _sibling_marker(target: Path) -> Path:
    return target.with_name(f"{target.name}{MARKER_SUFFIX}")


class HttpModelStore(ModelStore):
    """Download-with-resume model cache backed by an injected HTTP client."""

    def __init__(
        self,
        *,
        base_url: str,
        device_token: str,
        client: httpx.AsyncClient,
        cache_dir: Path | None = None,
        config: ModelCacheConfig | None = None,
        sleep: SleepFn = asyncio.sleep,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._client = client
        self._cache_dir = (cache_dir or DEFAULT_CACHE_DIR).expanduser()
        self._config = config or ModelCacheConfig()
        self._sleep = sleep
        self._auth_headers = {"Authorization": f"Bearer {device_token}"}
        self._locks: dict[str, asyncio.Lock] = {}

    # ── ModelStore interface ─────────────────────────────────────────────

    def path_if_present(self, manifest: ModelManifest) -> Path | None:
        """Return the blob path iff it (and any mmproj companion) exists and its
        marker matches the manifest sha256. Trusts the marker: does NOT rehash
        the file."""
        blob = blob_path(self._cache_dir, manifest)
        if not blob.exists():
            return None
        if read_marker(marker_path(self._cache_dir, manifest)) != manifest.sha256:
            return None
        if manifest.mmproj_file_name is not None:
            mmproj = blob.with_name(manifest.mmproj_file_name)
            if not mmproj.exists():
                return None
            if read_marker(_sibling_marker(mmproj)) != manifest.mmproj_sha256:
                return None
        return blob

    async def ensure(self, manifest: ModelManifest) -> Path:
        """Return a verified local path, downloading (with resume) if needed.

        A per-model lock serialises concurrent callers so a model is fetched at
        most once; the second caller re-checks presence under the lock.
        """
        present = self.path_if_present(manifest)
        if present is not None:
            return present
        async with self._lock_for(manifest.model_id):
            present = self.path_if_present(manifest)
            if present is not None:
                return present
            return await self._download_and_verify(manifest)

    # ── internals ────────────────────────────────────────────────────────

    def _lock_for(self, model_id: str) -> asyncio.Lock:
        # Safe without a guard: dict get/set has no await between them.
        lock = self._locks.get(model_id)
        if lock is None:
            lock = asyncio.Lock()
            self._locks[model_id] = lock
        return lock

    def _blob_url(self, model_id: str) -> str:
        return f"{self._base_url}{BLOB_PATH_TEMPLATE.format(model_id=model_id)}"

    def _mmproj_url(self, model_id: str) -> str:
        return f"{self._base_url}{MMPROJ_PATH_TEMPLATE.format(model_id=model_id)}"

    async def _download_and_verify(self, manifest: ModelManifest) -> Path:
        blob = await self._ensure_file(
            url=self._blob_url(manifest.model_id),
            target=blob_path(self._cache_dir, manifest),
            sha256=manifest.sha256,
            size_bytes=manifest.size_bytes,
        )
        if (
            manifest.mmproj_file_name is not None
            and manifest.mmproj_sha256 is not None
            and manifest.mmproj_size_bytes is not None
        ):
            await self._ensure_file(
                url=self._mmproj_url(manifest.model_id),
                target=blob.with_name(manifest.mmproj_file_name),
                sha256=manifest.mmproj_sha256,
                size_bytes=manifest.mmproj_size_bytes,
            )
        return blob

    async def _ensure_file(self, *, url: str, target: Path, sha256: str, size_bytes: int) -> Path:
        """Download-with-resume one file, verify it, and publish it atomically.

        A file whose marker already matches is returned as-is, so a manifest
        whose main blob verified but whose mmproj failed only refetches the
        mmproj on retry.
        """
        marker = _sibling_marker(target)
        if target.exists() and read_marker(marker) == sha256:
            return target
        part = target.with_name(f"{target.name}{PART_SUFFIX}")
        part.parent.mkdir(parents=True, exist_ok=True)
        result = await self._fetch_with_retries(url, part)
        self._verify_or_raise(target.name, sha256, size_bytes, part, result)
        write_marker_atomic(marker, sha256)
        part.replace(target)  # atomic publish of the verified file
        return target

    @staticmethod
    def _verify_or_raise(
        name: str, sha256: str, size_bytes: int, part: Path, result: DownloadResult
    ) -> None:
        if result.sha256 == sha256 and result.size == size_bytes:
            return
        part.unlink(missing_ok=True)
        raise ModelVerificationError(
            f"verification failed for {name}: "
            f"sha256 {result.sha256} vs {sha256}, "
            f"size {result.size} vs {size_bytes}"
        )

    async def _fetch_with_retries(self, url: str, part: Path) -> DownloadResult:
        attempt = 0
        while True:
            try:
                return await stream_to_part(
                    client=self._client,
                    url=url,
                    headers=self._auth_headers,
                    part=part,
                    chunk_size=self._config.chunk_size,
                )
            except (httpx.TransportError, RetryableStatusError) as exc:
                attempt += 1
                if attempt > self._config.max_retries:
                    raise ModelFetchError(
                        f"failed to fetch {url} after {attempt} attempt(s)"
                    ) from exc
                await self._sleep(self._config.backoff_base_s * (2 ** (attempt - 1)))

File: go-agent/modelcache/store.go
package modelcache

import (
	"context"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/Unluckyathecking/fallow/go-agent/protocol"
)

// Store is a download-with-resume model cache backed by an injected HTTP client.
// It implements the same contract as the Python HttpModelStore and shares its
// on-disk layout.
type Store struct {
	baseURL     string
	client      *http.Client
	cacheDir    string
	cfg         Config
	sleep       func(time.Duration)
	authHeaders map[string]string

	locksMu sync.Mutex
	locks   map[string]*sync.Mutex
}

// Option customises a Store at construction.
type Option func(*Store)

// WithCacheDir sets the cache root directory (default: ResolveDefaultCacheDir).
func WithCacheDir(dir string) Option { return func(s *Store) { s.cacheDir = dir } }

// WithConfig overrides the retry/backoff/chunk tuning (default: DefaultConfig).
func WithConfig(cfg Config) Option { return func(s *Store) { s.cfg = cfg } }

// WithSleep overrides the backoff sleep (default: time.Sleep). Tests use it to
// record backoff durations without waiting.
func WithSleep(sleep func(time.Duration)) Option { return func(s *Store) { s.sleep = sleep } }

// ResolveDefaultCacheDir expands the default cache root against the user home
// directory.
func ResolveDefaultCacheDir() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(home, DefaultCacheDir), nil
}

// New constructs a Store for the given coordinator base URL and device token.
func New(baseURL, deviceToken string, client *http.Client, opts ...Option) *Store {
	s := &Store{
		baseURL:     strings.TrimRight(baseURL, "/"),
		client:      client,
		cfg:         DefaultConfig(),
		sleep:       time.Sleep,
		authHeaders: map[string]string{"Authorization": "Bearer " + deviceToken},
		locks:       make(map[string]*sync.Mutex),
	}
	for _, opt := range opts {
		opt(s)
	}
	if s.cacheDir == "" {
		if dir, err := ResolveDefaultCacheDir(); err == nil {
			s.cacheDir = dir
		}
	}
	return s
}

// PathIfPresent returns the blob path and true iff the blob exists and its
// marker matches the manifest sha256. It trusts the marker and does NOT rehash
// the file, so it is safe to call on the heartbeat hot path.
func (s *Store) PathIfPresent(m protocol.ModelManifest) (string, bool) {
	blob := blobPath(s.cacheDir, m)
	if _, err := os.Stat(blob); err != nil {
		return "", false
	}
	if readMarker(markerPath(s.cacheDir, m)) != m.SHA256 {
		return "", false
	}
	if m.MmprojFileName != nil {
		mmproj := filepath.Join(modelDir(s.cacheDir, m), *m.MmprojFileName)
		if _, err := os.Stat(mmproj); err != nil {
			return "", false
		}
		if m.MmprojSHA256 == nil || readMarker(mmproj+MarkerSuffix) != *m.MmprojSHA256 {
			return "", false
		}
	}
	return blob, true
}

// Ensure returns a verified local path, downloading (with resume) if needed. A
// per-model lock serialises concurrent callers so a model is fetched at most
// once; the second caller re-checks presence under the lock.
func (s *Store) Ensure(ctx context.Context, m protocol.ModelManifest) (string, error) {
	if present, ok := s.PathIfPresent(m); ok {
		return present, nil
	}
	lock := s.lockFor(m.ModelID)
	lock.Lock()
	defer lock.Unlock()

	if present, ok := s.PathIfPresent(m); ok {
		return present, nil
	}
	return s.downloadAndVerify(ctx, m)
}

func (s *Store) lockFor(modelID string) *sync.Mutex {
	s.locksMu.Lock()
	defer s.locksMu.Unlock()
	lock := s.locks[modelID]
	if lock == nil {
		lock = &sync.Mutex{}
		s.locks[modelID] = lock
	}
	return lock
}

func (s *Store) blobURL(modelID string) string {
	return s.baseURL + fmt.Sprintf(BlobPathTemplate, modelID)
}

func (s *Store) mmprojURL(modelID string) string {
	return s.baseURL + fmt.Sprintf(MmprojPathTemplate, modelID)
}

func (s *Store) downloadAndVerify(ctx context.Context, m protocol.ModelManifest) (string, error) {
	blob, err := s.ensureFile(
		ctx, s.blobURL(m.ModelID), blobPath(s.cacheDir, m), m.SHA256, int64(m.SizeBytes),
	)
	if err != nil {
		return "", err
	}
	if m.MmprojFileName != nil && m.MmprojSHA256 != nil && m.MmprojSizeBytes != nil {
		target := filepath.Join(modelDir(s.cacheDir, m), *m.MmprojFileName)
		_, err := s.ensureFile(
			ctx, s.mmprojURL(m.ModelID), target, *m.MmprojSHA256, int64(*m.MmprojSizeBytes),
		)
		if err != nil {
			return "", err
		}
	}
	return blob, nil
}

// ensureFile downloads-with-resume one file, verifies it, and publishes it. A
// file whose marker already matches is returned as-is, so a manifest whose main
// blob verified but whose mmproj failed only refetches the mmproj on retry.
func (s *Store) ensureFile(
	ctx context.Context, url, target, sha256 string, size int64,
) (string, error) {
	marker := target + MarkerSuffix
	if _, err := os.Stat(target); err == nil && readMarker(marker) == sha256 {
		return target, nil
	}
	part := target + PartSuffix
	if err := os.MkdirAll(filepath.Dir(part), 0o755); err != nil {
		return "", err
	}
	result, err := s.fetchWithRetries(ctx, url, part)
	if err != nil {
		return "", err
	}
	if err := verifyOrDelete(filepath.Base(target), sha256, size, part, result); err != nil {
		return "", err
	}
	// Publish the verified bytes before the marker: a crash in between costs
	// one re-download, while the reverse order would leave a fresh marker
	// vouching for stale bytes that are then trusted without rehashing.
	if err := os.Rename(part, target); err != nil {
		return "", err
	}
	if err := writeMarkerAtomic(marker, sha256); err != nil {
		return "", err
	}
	return target, nil
}

// verifyOrDelete checks the downloaded bytes against the expected digest and
// size. On mismatch it deletes the partial file and returns ErrVerification.
func verifyOrDelete(name, sha256 string, size int64, part string, result downloadResult) error {
	if result.sha256 == sha256 && result.size == size {
		return nil
	}
	_ = os.Remove(part)
	return fmt.Errorf(
		"%w for %s: sha256 %s vs %s, size %d vs %d",
		ErrVerification, name, result.sha256, sha256, result.size, size,
	)
}

func (s *Store) fetchWithRetries(ctx context.Context, url, part string) (downloadResult, error) {
	attempt := 0
	for {
		result, err := streamToPart(ctx, s.client, url, s.authHeaders, part, s.cfg.ChunkSize)
		if err == nil {
			return result, nil
		}
		if !isRetryable(err) {
			return downloadResult{}, err
		}
		attempt++
		if attempt > s.cfg.MaxRetries {
			return downloadResult{}, fmt.Errorf(
				"%w: %s after %d attempt(s): %v", ErrFetch, url, attempt, err,
			)
		}
		s.sleep(s.cfg.BackoffBase * time.Duration(1<<(attempt-1)))
	}
}

// isRetryable reports whether an error from one download attempt should be
// retried: transport-level failures (surfaced by net/http as *url.Error) and
// retryable HTTP statuses. Content and filesystem errors are not retried.
//
// Context cancellation/deadline is checked FIRST: net/http wraps those in a
// *url.Error, so without this guard a cancelled download would spin through the
// full retry/backoff budget (~seconds of wasted work) instead of returning at
// once. A cancelled context will not become un-cancelled, so retrying is futile.
func isRetryable(err error) bool {
	if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
		return false
	}
	var status *retryableStatusError
	if errors.As(err, &status) {
		return true
	}
	var urlErr *url.Error
	return errors.As(err, &urlErr)
}
